cmd/gateway-azure.go

```diff
@@ -41,52 +41,112 @@ const globalAzureAPIVersion = "2016-05-31"
-// defined metadata to have X-Ms-Meta prefix.
-func s3ToAzureHeaders(headers map[string]string) (newHeaders map[string]string) {
+// s3MetaToAzure converts metadata meant for S3 PUT/COPY object into
+// Azure data structures - BlobMetadata and BlobProperties.
+//
+// BlobMetadata contains user defined key-value pairs and each key is
+// automatically prefixed with `X-Ms-Meta-` by the Azure SDK. S3
+// user-metadata is translated to Azure metadata by removing the
+// `X-Amz-Meta-` prefix.
+//
+// BlobProperties contains commonly set metadata for objects such as
+// Content-Encoding, etc. Such metadata that is accepted by S3 is
+// copied into BlobProperties.
+//
+// Header names are canonicalized as in http.Header.
+func s3MetaToAzure(metadata map[string]string) (storage.BlobMetadata, storage.BlobProperties) {
+	// Azure does not permit user-defined metadata key names to
+	// contain hyphens. So we map hyphens to underscores for
+	// encryption headers. More such headers may be added in the
+	// future.
 	gatewayHeaders := map[string]string{
 		"X-Amz-Meta-X-Amz-Key":     "X-Amz-Meta-x_minio_key",
 		"X-Amz-Meta-X-Amz-Matdesc": "X-Amz-Meta-x_minio_matdesc",
 		"X-Amz-Meta-X-Amz-Iv":      "X-Amz-Meta-x_minio_iv",
 	}
 
-	newHeaders = make(map[string]string)
-	for k, v := range headers {
+	blobMeta := make(map[string]string)
+	var props storage.BlobProperties
+	for k, v := range metadata {
 		k = http.CanonicalHeaderKey(k)
 		if nk, ok := gatewayHeaders[k]; ok {
 			k = nk
 		}
-		if strings.HasPrefix(k, "X-Amz-Meta") {
-			k = strings.Replace(k, "X-Amz-Meta", "X-Ms-Meta", -1)
+		switch {
+		case strings.HasPrefix(k, "X-Amz-Meta-"):
+			// Strip header prefix, to let Azure SDK
+			// handle it for storage.
+			k = strings.Replace(k, "X-Amz-Meta-", "", 1)
+			blobMeta[k] = v
+
+		// All cases below, extract common metadata that is
+		// accepted by S3 into BlobProperties for setting on
+		// Azure - see
+		// https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html
+		case k == "Cache-Control":
+			props.CacheControl = v
+		case k == "Content-Disposition":
+			props.ContentDisposition = v
+		case k == "Content-Encoding":
+			props.ContentEncoding = v
+		case k == "Content-Length":
+			// assume this doesn't fail
+			props.ContentLength, _ = strconv.ParseInt(v, 10, 64)
+		case k == "Content-MD5":
+			props.ContentMD5 = v
+		case k == "Content-Type":
+			props.ContentType = v
 		}
-		newHeaders[k] = v
 	}
-	return newHeaders
+	return blobMeta, props
 }
 
-// Prefix user metadata with "X-Amz-Meta-".
-// client.GetBlobMetadata() already strips "X-Ms-Meta-"
-func azureToS3Metadata(meta map[string]string) (newMeta map[string]string) {
+// azureToS3Meta converts Azure metadata/properties to S3 metadata. It
+// is the reverse of s3MetaToAzure. Azure's `.GetMetadata()`
+// lower-cases all header keys, so this is taken into account by this
+// function.
+func azureToS3Meta(meta storage.BlobMetadata, props storage.BlobProperties) map[string]string {
```

Comment:
you could rename this function as `azurePropertiesToS3Meta()`

Reply:
Done.